### PR TITLE
Fix template migration issues

### PR DESCRIPTION
**Pull request by philipp-spiess**
This PR fixes two issues we found when testing the candidate codemodes:

1. Sometimes, core would emit the same candidate twice. This would result into rewriting a range multiple times, without realizing that this change might already be applied, causing it to swallow or duplicate some bytes.
2. The codemods were mutating the `Candidate` object, however since the `Candidate` parsing is _cached_ in core, it would sometimes return the same instance. This is an issue especially since we monkey patch the prefix to `null` when migrating prefixed candidates. This means that a candidate would be cached that would be _invalid relative to the real design system_. We fixed this by making sure the mutations would only be applied to clones of the `Candidate` and I changed the `DesignSystem` API to return `ReadOnly<T>` versions of these candidates. A better solution would maybe be to disable the cache at all but this requires broader changes in Core.

### CODE REVIEW

File: packages/@tailwindcss-upgrade/src/template/candidates.ts [hunk 1]
```diff
@@ -5,15 +5,24 @@ import type { DesignSystem } from '../../../tailwindcss/src/design-system'
 
 export async function extractRawCandidates(
   content: string,
+  extension: string,
 ): Promise<{ rawCandidate: string; start: number; end: number }[]> {
   let scanner = new Scanner({})
-  let result = scanner.getCandidatesWithPositions({ content, extension: 'html' })
+  let result = scanner.getCandidatesWithPositions({ content, extension })
```
Comment (philipp-spiess) [inline packages/@tailwindcss-upgrade/src/template/candidates.ts:11]: Note to myself: Let's try to see if we can find out why this emits duplicates

File: packages/@tailwindcss-upgrade/src/template/candidates.ts [hunk 2]
```diff
@@ -5,15 +5,24 @@ import type { DesignSystem } from '../../../tailwindcss/src/design-system'
...
+  let candidatesMap = new Map<string, { rawCandidate: string; start: number; end: number }>()
 
-  let candidates: { rawCandidate: string; start: number; end: number }[] = []
   for (let { candidate: rawCandidate, position: start } of result) {
-    candidates.push({ rawCandidate, start, end: start + rawCandidate.length })
+    let end = start + rawCandidate.length
+    candidatesMap.set(`${start}:${end}:${rawCandidate}`, {
+      rawCandidate,
+      start,
+      end: start + rawCandidate.length,
```
Comment (adamwathan) [inline packages/@tailwindcss-upgrade/src/template/candidates.ts:21]: Can just use the `end` variable here it looks like 👍

File: packages/@tailwindcss-upgrade/src/template/candidates.ts (hunk 1 above)
Comment (philipp-spiess) [reply packages/@tailwindcss-upgrade/src/template/candidates.ts:11]: Wasn't really able to repro this in unit tests 🤔

Comment (thecrypticace) [reply packages/@tailwindcss-upgrade/src/template/candidates.ts:11]: We can chat about this some time today but I'd guess one of the routines we use to slice up a candidate into multiple candidates produces two identical entries.

Comment (philipp-spiess) [reply packages/@tailwindcss-upgrade/src/template/candidates.ts:11]: @thecrypticace Sounds good yeah, I was trying to reproduce it inside Rust unit tests to no avail. I wonder if this has to do with the fact that we have two files in the repro—I have a feeling that there's some shared state but I didn't see anything that stood out.

**Review by adamwathan — APPROVED** [review]

File: CHANGELOG.md [hunk 3]
```diff
@@ -18,6 +18,7 @@ and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0
 - Don’t crash when scanning a candidate equal to the configured prefix ([#14588](https://github.com/tailwindlabs/tailwindcss/pull/14588))
 - Ensure there's always a space before `!important` when stringifying CSS ([#14611](https://github.com/tailwindlabs/tailwindcss/pull/14611))
 - _Experimental_: Ensure CSS before a layer stays unlayered when running codemods ([#14596](https://github.com/tailwindlabs/tailwindcss/pull/14596))
+- _Experimental_: Resolve issues where some prefixed candidates were not properly migrated ([#14600](https://github.com/tailwindlabs/tailwindcss/pull/14600))
```
Comment (philipp-spiess) [inline CHANGELOG.md:21]:
```suggestion
- _Upgrade (experimental)_: Ensure CSS before a layer stays unlayered when running codemods ([#14596](https://github.com/tailwindlabs/tailwindcss/pull/14596))
- _Upgrade (experimental)_: Resolve issues where some prefixed candidates were not properly migrated ([#14600](https://github.com/tailwindlabs/tailwindcss/pull/14600))
```

**Pull request merged**